### ISSUE TITLE
godef doesn't work with source files containing windows line endings and characters in utf-8

### ISSUE THREAD

**Issue by evgenybf** [ironcladlou/GoTools#112]
It turned out that view.text_point returns the same offset for files which contain Linux or Windows line endings. That's wrong as windows line endings are composed of two characters: 0x0d 0x0a, while linux line endings are just a single character 0x0a. As a result, godef (and others!) can't find correct identifier in a text when it's in the Windows format.

I hope that there is better workaround than converting all Go files to linux format...  For instance
```python
  @staticmethod
  def location_at_cursor(view):
    row, col = view.rowcol(view.sel()[0].begin())
    offsets = Buffers.offset_at_cursor(view)
    o0, o1 = offsets[0], offsets[1]
    if view.line_endings() == "Windows":
        o0 += row
        o1 += row
    return (view.file_name(), row, col, o0, o1)
```
but there is still a problem with mixed line endings though.

PS: Sublime's menu for manipulating the endings: View/Line Endings

**Comment by ironcladlou** [comment]
Sounds like a bug in `sublime.View.text_point` for sure. The workaround you proposed seems like it'd be better than nothing. Without the patch, it's wrong most of the time, and with the patch it'll only be wrong some of the time (and I'm not really concerned about mixed line endings- I consider that a developer problem :grin:).

Want to submit a PR?

**Comment by ironcladlou** [comment]
We should probably come up with a simple test scenario and file a bug upstream at https://github.com/SublimeTextIssues/Core.

**Comment by evgenybf** [comment]
@ironcladlou, It's possible that result of text_point is not supposed to be used outside of Sublime. I assume that Sublime cut off all line endings when it loads a file into memory.

**Comment by ironcladlou** [comment]
@evgenybf 
>It's possible that result of text_point is not supposed to be used outside of Sublime. I assume that Sublime cut off all line endings when it loads a file into memory.

That's a fair point. I guess the only way to know for sure will be to engage upstream. If the offset isn't supposed to and will never work the way we need, I'm not sure what else to do at the moment.

**Comment by evgenybf** [comment]
https://github.com/SublimeTextIssues/Core/issues/989

**Comment by evgenybf** [comment]
The same problem with files with utf-8 characters.

https://gist.github.com/evgenybf/f1436c5daf0c0fcac975

**Comment by ironcladlou** [comment]
Looks like we're going to need some novel way to get a correct byte offset portably and efficiently. I have no ideas at the moment.

**Comment by evgenybf** [comment]
Quick workaround based on 
https://github.com/clemos/haxe-sublime-bundle/blob/master/features/haxe_find_definition.py
```python
class Buffers():
....
  @staticmethod 
  def get_offset(view, pt, row):
    src = view.substr(sublime.Region(0, pt))
    import codecs
    offset = len(codecs.encode(src, "utf-8")) + 1 # add 1 because offset is 1-based
    if view.line_endings() == "Windows":
       offset += row
    return offset

  @staticmethod
  def offset_at_cursor(view):
    begin_row, begin_col = view.rowcol(view.sel()[0].begin())
    end_row, end_col = view.rowcol(view.sel()[0].end())
    return (Buffers.get_offset(view, view.sel()[0].begin(), begin_row), 
        Buffers.get_offset(view, view.sel()[0].end(), end_row))
......
  @staticmethod
  def location_for_event(view, event):
    pt = view.window_to_text((event["x"], event["y"]))
    row, col = view.rowcol(pt)
    offset = Buffers.get_offset(view, pt, row)
    return (view.file_name(), row, col, offset)
```
Please, do not use it as is. It's just a draft without any optimizations.

**Comment by ironcladlou** [comment]
Interesting approach... I'll tinker with it. Thanks, good find!

**Comment by ironcladlou** [comment]
This exact scenario and discussion happened in 2011 on vim-dev... Pretty interesting just how similar it is. https://groups.google.com/forum/#!topic/vim_dev/7XLbDYHXCfA

**Comment by stevekuznetsov** [comment]
@ironcladlou Really funny how their encodings screw up each others' names in the quoted text they're responding to.

**Comment by liuhewei** [comment]
@ironcladlou Could you take a look at the implementation of GoOracle: https://github.com/waigani/GoOracle/blob/master/goOracle.py#L25

It might help.

**Comment by ironcladlou** [comment]
@evgenybf @liuhewei would you be willing to test the fix in #125? I've done a few simple tests locally and I don't notice any obvious regressions. I don't have the time right now to set up a test which would exercise Windows line endings, etc.
